fabric8a_auth/auth.py

```diff
@@ -0,0 +1,208 @@
+    """Get the authorization token read from the request header."""
+    return request.headers.get('Authorization')
+
+
+def get_audiences():
+    """Retrieve all JWT audiences."""
+    return os.environ.get('FABRIC8_ANALYTICS_JWT_AUDIENCE', '').split(',')
+
+
+def login_required(view):
+    """Check if the login is required and if the user can be authorized."""
+    @wraps(view)
+    def wrapper(*args, **kwargs):
+        if os.environ.get('DISABLE_AUTHENTICATION') in ('1', 'True', 'true'):
+            return view(*args, **kwargs)
+
+        lgr = current_app.logger
+
+        try:
+            decoded = decode_user_token(current_app, get_token_from_auth_header())
+            if not decoded:
+                lgr.exception('Provide an Authorization token with the API request')
+                raise AuthError(401, 'Authentication failed - token missing')
+
+            lgr.info('Successfully authenticated user {e} using JWT'.
+                     format(e=decoded.get('email')))
+        except jwt.ExpiredSignatureError as exc:
+            lgr.exception('Expired JWT token')
+            raise AuthError(401, 'Authentication failed - token has expired') from exc
+        except Exception as exc:
+            lgr.exception('Failed decoding JWT token')
+            raise AuthError(401, 'Authentication failed - could not decode JWT token') from exc
+
+        return view(*args, **kwargs)
+
+    return wrapper
+
+
+def service_token_required(view):
+    """Check if the request contains a valid service token."""
+    @wraps(view)
+    def wrapper(*args, **kwargs):
+        if os.environ.get('DISABLE_AUTHENTICATION') in ('1', 'True', 'true'):
+            return view(*args, **kwargs)
+
+        lgr = current_app.logger
+
+        try:
+            decoded = decode_service_token(get_token_from_auth_header())
+            if not decoded:
+                lgr.exception('Provide an Authorization token with the API request')
+                raise AuthError(401, 'Authentication failed - token missing')
+
+            lgr.info('Successfully authenticated user {e} using JWT'.
+                     format(e=decoded.get('email')))
+        except jwt.ExpiredSignatureError as exc:
+            lgr.exception('Expired JWT token')
+            raise AuthError(401, 'Authentication failed - token has expired') from exc
+        except Exception as exc:
+            lgr.exception('Failed decoding JWT token')
+            raise AuthError(401, 'Authentication failed - could not decode JWT token') from exc
+
+        return view(*args, **kwargs)
+
+    return wrapper
+
+
+def fetch_public_keys(app):
+    """Get public keys for OSIO service account. Currently, there are three public keys."""
+    if not getattr(app, "public_keys", []):
+        auth_url = os.environ.get('OSIO_AUTH_URL')
+        if auth_url:
+            try:
+                auth_url = auth_url.strip('/') + '/api/token/keys?format=pem'
+                result = requests.get(auth_url, timeout=2)
```

Comment:
not sure whether we want to put timeout into configurable env.variable?

Reply:
Yep, it makes.